### ISSUE TITLE
How to implement "array label" input via ops.ExternalSource()

### ISSUE THREAD

**Issue by wangguangyuan** [NVIDIA/DALI#123]
https://github.com/NVIDIA/DALI/blob/master/dali/benchmark/resnet50_bench.py
The example given is to read all the pictures into the memory. When our data is very large, it is impossible to read all the data into the memory. We need to read only the batch_size image at a time. At the same time, the example passes the image through ops.ExternalSource(), but does not pass the label corresponding to the image, which causes me to match the image and the label when I can't retrain. Is there any way to pass the image and label together via ops.ExternalSource()?
At the same time, I am puzzled that we can achieve the multi-threaded processing by rewriting iter_setup(self) to send external data？

**Comment by JanuszL** [comment]
Hi,
That is a good question. In this example, we load all the data at the beginning as the input set is small. In the real use case, you should put your loading code inside _iter_setup_ and load batch of images at each iteration. Regarding loading the labels, you can treat them as any other data and load them using _ExternalSource_ as well. In such case, you will have one source for images and the second one for the labels.
> At the same time, I am puzzled that we can achieve the multi-threaded processing by rewriting iter_setup(self) to send external data？

If I understand your question correctly, you can multi-threaded processing inside _iter_setup_ but this code needs to feed data using _feed_input_ to all _ExternalSources_ before _iter_setup_ ends. However, you can continue you background processing outside _iter_setup_.
@ptrendx I hope I haven't missed anything.

**Comment by JanuszL** [comment]
Registered at DALI-207

**Comment by wangguangyuan** [comment]
The function used in the dali serialization process must be provided by dali?
Is it not possible to serialize an external function?
If so, is there any other solution?

**Comment by JanuszL** [comment]
Hi,
Currently external source is not serialized and deosn't work for TF. Real reason is that DALI cannot call python code from inside - in this case iter_setup is not callable (with current design) and there is no easy way to feed data into ExternalSource. It should work with PyTorch as well as MxNet.
We plan to develop such functionality in the future but it is hard to tell now when.

**Comment by wangguangyuan** [comment]
Hi, thank you very much, is there a document about the dali architecture? Some of the features I need may need to be added from source.

**Comment by JanuszL** [comment]
Hi,
Please look into https://github.com/NVIDIA/DALI#additional-resources first. Regarding detailed documentation which describes a design of each class, we don't have such document and it is probably not going to be done anytime soon. You can still use Doxygen file from DALI source to generate development docs.

**Comment by klecki** [comment]
Hi,
the support for External Source callbacks/iterator in TensorFlow via the tf.data.Dataset compatible API was merged and will be present in DALI 1.5.

DALI 1.4 already supports inputs from other tf.data.Datasets in the experimental.DALIDatasetWithInputs.

You can see more in the documentation: https://docs.nvidia.com/deeplearning/dali/main-user-guide/docs/plugins/tensorflow_plugin_api.html#experimental

Tutorial is under review in: https://github.com/NVIDIA/DALI/pull/3212

**Comment by JanuszL** [comment]
Hi,
DALI 1.5 is out and can be used to test the new functionality.
